### DISCUSSION TITLE
Rectangular selection by mouse

### DISCUSSION THREAD

**Discussion by dov** [vsg-dev/VulkanSceneGraph#1298] [General]
I'm trying to create a rectangular selection by the mouse. Meanwhile I have managed to draw my selection rectangle as a separate `vsg::View` and have it properly rendered before I enter the vsg `advanceToNextFrame()` main loop. But when I try to update the rectangle position (actually I'm deleting the old rectangle, and creating a new one inside of the loop) I encounter two problems:

1. The color of the rectangle doesn't change, and is black. Even though I was able to control it in the pre-mainloop node.
2. I get the following error from the Vulkan debug Layers:

```
    Objects: 0
VUID-vkDestroyBuffer-buffer-00922(ERROR / SPEC): msgNum: -464217071 - Validation Error: [ VUID-vkDestroyBuffer-buffer-00922 ] | MessageID = 0xe4549c11 | vkDestroyBuffer():  can't be called on VkBuffer 0x5e8ea90000000359[] that is currently in use by VkDescriptorSet 0x59ffe0000000003d[]. The Vulkan spec states: All submitted commands that refer to buffer, either directly or via a VkBufferView, must have completed execution (https://vulkan.lunarg.com/doc/view/1.3.283.0/linux/1.3-extensions/vkspec.html#VUID-vkDestroyBuffer-buffer-00922)
```

What am I doing wrong?

Here is my main loop with the rectangle update section:
```c++
  while (viewer->advanceToNextFrame())
  {
    // pass any events into EventHandlers assigned to the Viewer
    viewer->handleEvents();

    // update the mouse drawn rectangle
    if (mouseHandler->isPressed && !mouseDown)
    {
      mouseDown = true;
      mouseDownPos = vsg::vec2(mouseHandler->lastPointerEvent->x, mouseHandler->lastPointerEvent->y);
    }
    else if (!mouseHandler->isPressed && mouseDown)
      mouseDown = false;
    if (mouseDown)
    {
      auto mousePos = vsg::vec2(mouseHandler->lastPointerEvent->x, mouseHandler->lastPointerEvent->y);
      auto rect = Rectangle {mouseDownPos, mousePos - mouseDownPos};
      rect.pos.y = window->extent2D().height - rect.pos.y;
      rect.size.y = -rect.size.y;
      if (rect.size.x < 0)
      {
        rect.pos.x += rect.size.x;
        rect.size.x = -rect.size.x;
      }
      if (rect.size.y < 0)
      {
        rect.pos.y += rect.size.y;
        rect.size.y = -rect.size.y;
      }
      view_overlay->children.clear();
      view_overlay->addChild(draw_rectangle(window, rect));
      viewer->compile();
    }

    viewer->update();

    viewer->recordAndSubmit();

    viewer->present();

    numFramesCompleted += 1.0;
  }
```

The complete `select-rectangle.cpp` example is here:

https://gist.github.com/dov/90763c1138e0db214430e1d4c2325cff

While, I'm at it, I'd love to get some tips of how I can intersect the rectangle with the scene and find either of the following:

1. A list of meshes that partially intersect the selection rectangle
2. A list of meshes that are fully enclosed by the selection rectangle

Thanks in advance!

**Comment by jacky-wang7** [comment]
I guess there is a problem with children release.
```
view_overlay->children.clear();
```
this discussion may help you:
https://github.com/vsg-dev/VulkanSceneGraph/discussions/1273

Regarding the intersection of the rectangles, is this what you were looking for?
https://github.com/vsg-dev/VulkanSceneGraph/discussions/1288

**Reply by jacky-wang7** [reply]
For drawing a rectangle, perhaps you can update the rectangle by updating the vertex coordinates instead of creating a new node.
I'm not an expert, just some idea.

**Reply by dov** [reply]
Thanks @robertosfield  and @jacky-wang7 . That should give me enouch info to implement it.

**Comment by robertosfield** [comment]
Rather than create the rectangle on the fly what I would recommend is create a fixed size rectangle at start up and scale and place it using a vsg::MatrixTransform, updating the transforms matrix each time I want to position/size it.  Putting a vsg::Switch above the subgraph if I want to toggle it on/off.

For picking the objects in the scene using the extents of the rectangle the tool to use is the new vsg::PolytopeIntersector that I check into VSG master last week.  There's a option for section up the polytope based on window coords, the vsgintersection example now has a code path for PolytopeIntersector so have a look at vsgExamples master for this addition.

**Reply by dov** [reply]
Thanks, @robertosfield . I have done exactly that, and my sample code is working (almost) great. 

The remaining problem is that once I resize the window, the rectangle is no longer aligned with the mouse pointer.

I realized that the third parameter in the creation of a `vsg::Camera` is the `ViewPort`, which for my example is initialized with the window size, as extracted  by `vsg::Window::extent2D()`. When the window size changes, I therefore tried to update the `ViewPort` to match the new window size as follows:
  
 ```c++
  void apply(vsg::ConfigureWindowEvent& confEvent) override
  {
    _camera->viewportState->set(0,0,confEvent.width, confEvent.height);
  }
```
  
I tried doing this in my updated select-rectangle.cpp version, see gist below, but it didn't work. I'm getting the wrong coordinates for my transformed quad (rectangle), after resizing the window, even though I updated the camera viewportstate. as above.

I also tried doing `viewer->compile` after updating the `viewPortState`, but it didn't help either. Ι verified that if I initally create the window e.g. twice as big in x, then the rectangle does follow the mouse.

So what went wrong?

The updated version can be found here: 

https://gist.github.com/dov/77cf7d66fdb8cacce897dc8c4f893f7c

Thank you very much in advance!

**Reply by dov** [reply]
I found the reason, and a workaround, though I still don't understand all the steps.

When a window is resized, `camera.projectionMatrix` is updated "automatically". In my 1:1 pixel mapping I had assumed that it would be stay at the constant (0,-1, 0,-1,  -1,1) as I had initialized it, but after resizing the window, "someone", and I guess it has to do with `ViewDependentState`, updates it to compensate for the change of size. Since I'm doing the scaling through the `viewport`, these changes interacted destructively. By resetting the `projectionMatrix` to my desired value in the update callbacks, was I able to render the rectangle correctly.

**Reply by jacky-wang7** [reply]
> 我找到了原因和解决方法，尽管我仍然不理解所有步骤。
> 
> 调整窗口大小时，将“自动”更新。在我的 1：1 像素映射中，我假设它会保持常数 （0，-1， 0，-1， -1,1），因为我已经初始化了它，但在调整窗口大小后，“某人”，我猜这与 有关，更新了它以补偿大小的变化。由于我是通过 进行扩展的，因此这些更改会以破坏性方式交互。通过在 update 回调中将 the 重置为我想要的值，我是否能够正确渲染矩形。`camera.projectionMatrix``ViewDependentState``viewport``projectionMatrix`

WindowResizeHandler will change the camera state when the window is resized.So you may need to set the camera matrix yourself after the WindowResizeHandler , or consider disabling the windowResizeHandler. This may be the problem you are encountering, some ideas.

**Reply by dov** [reply]
Thanks @jacky-wang7 . That indeed explains it! I verified that I can indeed control the the camera by overriding the handler. 

However, as the `renderGraph` may contain several views that share the same `windowResizeHandler`, it gets a bit messy to do the override selectively, and I'll just keep my current approach of overriding the projectionMatrix right before I update the matrix transformations of the selection rectangle. A bit redundant, but it practice it doesn't matter.

**Comment by robertosfield** [comment]
I'm sorry, I don't have the time available to investigate what you are doing wrong.

**Reply by dov** [reply]
Thanks @robertosfield . I totaly understand, and in any case I did find a perfect work around (though I still wonder why the `camera.projectionMatrix` was changed, and how I can prevent it?).

Don't get me wrong, I'm in awe, and very appreciative of your great gift to the world that is VSG. By opening issues about things that I got stuck with during my journey of learning it, I can only hope that you or someone else, can share some insight that may lead me closer to the solution. I know open source well enough that it is obvious to me, that nobody "owes" anybody anything, but still, by humbly describing issues in detail, sometimes the story may awaken the professional curiosity of someone like you, who may or may not also have the time to answer.

So once again, thanks. And there is no need to ever feel pressured to answer.

**Reply by robertosfield** [reply]
Unfortunately I have a big TODO list so I have to prioritize.  Most of my time in past 2 months has been tackle items on the TODO list that are important open source items that are unpaid, thankfully I've earned enough in the past year to make this possible, but still I do need to balance this work load and do some paid work to pay the bills.  I can't keep adding more unpaid items to my TODO list, so which I might have that professional curiosity and want to help, I have focus on the most critical unpaid items for the whole project rather than individual users.
